### DISCUSSION TITLE
search time

### DISCUSSION THREAD

**Discussion by ali-apadana** [bagisto/bagisto#8537] [General]
Hello every one,
I tested bagisto with 4,000 products. search section took about 0.7 seconds to fetch data. Is it normal?
I have about 3 million products and each one has about 3-4 variants.  How long does it take to get them?
(my server has 128 GB ram)

**Comment by devansh-webkul** [comment]
Hello there,

We appreciate your engagement with Bagisto and would like to assist you in optimizing its performance for your specific requirements. To better understand your situation and provide accurate guidance, we kindly request the following information:

**Bagisto Version**: Please specify the version of Bagisto you are currently using. You can find this information in your Bagisto admin dashboard or within the application files.

**Elasticsearch Integration**: Are you utilizing Elasticsearch for search functionality in your Bagisto?

Thanks!

**Reply by ali-apadana** [reply]
hi

I'm using Version : v2.0.0-BETA-1 

I enabled elasticsearch in bagisto's admin panel. but i got "No alive nodes found in your cluster" error.

I installed elasticsearch on my server but I can't start it! I don't know why!
sudo service elasticsearch start
Starting elasticsearch (via systemctl):  Job for elasticsearch.service failed because the control process exited with error code.
See "systemctl status elasticsearch.service" and "journalctl -xe" for details.
                                                           [FAILED]

can u help me please?

**Reply by devansh-webkul** [reply]
Looks like your Elasticsearch is not properly installed.

**Reply by ali-apadana** [reply]
finally it has been installed:))
it works in console with this command : "curl -X GET "http://localhost:9200" -u elastic:PASSWORD"
I set these in bagisto .env :
ELASTICSEARCH_USER=elastic
ELASTICSEARCH_PASS=PASSWORD

But I'm getting error

{"error":{"root_cause":[{"type":"security_exception","reason":"missing authentication credentials for REST request [/products_default_en_index/_search]","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","ApiKey"]}}],"type":"security_exception","reason":"missing authentication credentials for REST request [/products_default_en_index/_search]","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","ApiKey"]}},"status":401}

**Reply by ali-apadana** [reply]
fixed by "php artisan config:cache"

Now This is error!
{"error":{"root_cause":[{"type":"query_shard_exception","reason":"No mapping found for [price] in order to sort on","index_uuid":"Rak_HDpgSFOKP7t2d_w7xA","index":"products_default_en_index"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":0,"index":"products_default_en_index","node":"B6Q5UVXHTVSTd60ZUqz5Iw","reason":{"type":"query_shard_exception","reason":"No mapping found for [price] in order to sort on","index_uuid":"Rak_HDpgSFOKP7t2d_w7xA","index":"products_default_en_index"}}]},"status":400}

**Reply by ali-apadana** [reply]
In ElasticSearchRepository I Changed These:

in method getSortOptions:
`return [
            $sort => [
                'order' => $options['order'],
                **'unmapped_type' => 'long'**
            ],
        ];`
And in method search:
`$results = Elasticsearch::search([
            'index' => $this->getIndexName(),
            'body'  => [
                'from'          => $options['from'],
                'size'          => $options['limit'],
                **'track_total_hits' =>  true,**
                'stored_fields' => [],
                'query'         => [
                    'bool' => $filters ?: new \stdClass(),
                ],
                'sort'          => $this->getSortOptions($options),
            ],
        ]);`


Result is awesome!
Getting results with 15,000 products takes just 0.02 secs.
Thanks Elastic :)

And one more question about this is: Is Elasticsearch completely Free? or this is free trial?

**Reply by devansh-webkul** [reply]
Yes, it's free.
